### PR TITLE
Add environment to the relevant browse pages

### PR DESCRIPTION
**Pull request by jwforres**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1278579

### CODE REVIEW

**Comment by jwforres** [comment]
environment_collapse

After clicking "See all" on a long line it expands, and pretty prints if its JSON

environment-expanded

When a container has no env specified:

environment-none

**Comment by jwforres** [comment]
Not ready for code review yet, working through the design.  Currently adds an Environment tab, in theory this would be on Pods, Builds, anywhere else?   Another option would be to have the environment in the details tab, but that tab is already getting cluttered and env sometimes takes a lot of space.

@spadgett @smarterclayton @bparees @ajacobs21e

**Comment by smarterclayton** [comment]
We should put it on anything with a pod template, and both builds and
build configs. Build config and build will definitely share visual
editing elements.  Ultimately someone needs to be able to identify
common "shapes" - ie deployments, pods, rc, daemon sets, pet sets, etc
and have familiar editing on them.

**Comment by smarterclayton** [comment]
This is more visual and can be sorted out later, but the table borders and cell padding break up the intent of this section and draw the eye more than I would expect.  I also suspect that mono space may be more appropriate for the keys - I'm not sure this presentation favors the connection a developer would have to environment.

**Comment by jwforres** [comment]
I went with what we did for annotations for consistency.  If you think
environment deserves a different visual treatment I can take a different
approach.

**Comment by smarterclayton** [comment]
It's something we can sort out later.

**Comment by jwforres** [comment]
@spadgett ready for an initial review

File: assets/app/scripts/directives/truncate.js [hunk 1]
```diff
@@ -3,34 +3,21 @@
...
-            // Find the last word break, but don't look more than 10 characters back.
-            // Make sure we show at least the first 5 characters.
-            var startIndex = Math.max(4, scope.limit - 10);
-            var lastSpace = scope.visibleContent.lastIndexOf(' ', startIndex);
-            if (lastSpace !== -1) {
-              scope.visibleContent = scope.visibleContent.substring(0, lastSpace);
-            }
+          scope.visibleContent = truncateFilter(content, scope.limit, scope.useWordBoundary, scope.newlineLimit);
+          if (scope.visibleContent.length !== content.length) {
+            scope.truncated = true;
```
Comment (spadgett) [inline assets/app/scripts/directives/truncate.js]: If the content changes to something short, this is never reset to false. Maybe just

```
scope.truncated = scope.visibleContent.length !== content.length;
```

or you could even just check that in the view and get rid of the scope var.

File: assets/app/views/browse/deployment-config.html [hunk 2]
```diff
@@ -48,76 +48,89 @@
...
-                    <dd>Config</dd>
-                  </span>
-                </span>
-              </div>
-            </dl>
+        </uib-tab>
+        <uib-tab heading="Environment" active="selectedTab.environment">
+          <uib-tab-heading>Environment</uib-tab-heading>
+          <div ng-repeat="container in deploymentConfig.spec.template.spec.containers">
+            <h3>Container {{container.name}}</h3>
```
Comment (spadgett) [inline assets/app/views/browse/deployment-config.html]: Is there enough padding above each heading when there are multiple containers?

File: assets/app/views/directives/environment.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<div ng-if="envVars.length" class="table-responsive" style="margin-top: 5px;">
```
Comment (spadgett) [inline assets/app/views/directives/environment.html]: Probably better to switch to a stacked representation on mobile (name above value), although that might not be simple using a table.

File: assets/app/views/directives/environment.html [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  <table class="table table-bordered environment-variables">
+    <tbody>
+      <tr ng-repeat="env in envVars">
+        <td>{{env.name}}</td>
+        <td>
+          <!-- TODO we need to handle the valueFrom when it is defined instead of value -->
+          <span ng-if="!expanded[env.name] && (env.value | truncate : 200 : true : 3).length !== env.value.length">
+            <span class="env-var-value"><truncate-long-text content="env.value" limit="200" newline-limit="3"></truncate-long-text></span>
```
Comment (spadgett) [inline assets/app/views/directives/environment.html]: Would this logic be simpler if we included an optional see all link in the truncate-long-text directive?

File: assets/app/views/directives/environment.html [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<div ng-if="envVars.length" class="table-responsive" style="margin-top: 5px;">
+  <table class="table table-bordered environment-variables">
+    <tbody>
+      <tr ng-repeat="env in envVars">
```
Comment (spadgett) [inline assets/app/views/directives/environment.html]: We should have at least hidden table headers for accessibility.

**Comment by spadgett** [comment]
I like the expand/collapse behavior. It would be nice to reuse that for annotations.

File: assets/app/views/directives/environment.html (hunk 3 above)
Comment (spadgett) [inline assets/app/views/directives/environment.html]: > Probably better to switch to a stacked representation on mobile (name above value), although that might not be simple using a table.

Maybe a follow on.

**Comment by jwforres** [comment]
rebased, going to start looking at review comments now

File: assets/app/scripts/filters/util.js [hunk 6]
```diff
@@ -405,4 +405,30 @@ angular.module('openshiftConsole')
...
+      if (charLimit) {
+        truncated = truncated.substring(0, charLimit);
+      }
+
+      if (newlineLimit) {
+        var nthNewline = str.split("\n", newlineLimit).join("\n").length;
+        truncated = truncated.substring(0, nthNewline);
+      }
+
+      if (useWordBoundary !== false) {
```
Comment (spadgett) [inline assets/app/scripts/filters/util.js]: I suggest we flip the logic if we want to use word boundaries by default. Maybe

```
function(str, charLimit, ignoreWordBoundary, newlineLimit)
```

and

```
if (!ignoreWordBoundary) { ... }
```

**Comment by spadgett** [comment]
Just one comment, otherwise LGTM. Glad we moved some of the behavior into the truncate long text directive so we can reuse it.

**Comment by jwforres** [comment]
[merge]

**Comment by spadgett** [comment]
[merge]

**Pull request merged**